pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -748,3 +752,152 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
 func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
 	return nil, false
 }
+func (os *OpenStack) AttachDisk(diskName string) (string, error) {
+	disk, err := os.getVolume(diskName)
+	if err != nil {
+		return "", err
+	}
+	cClient, err := openstack.NewComputeV2(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	if err != nil || cClient == nil {
+		glog.Errorf("Unable to initialize nova client for region: %s", os.region)
+		return "", err
+	}
+
+	compute_id, err := os.getComputeIDbyHostname(cClient)
+	if err != nil || len(compute_id) == 0 {
+		glog.Errorf("Unable to get minion's id by minion's hostname")
+		return "", err
+	}
+
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil {
+		if compute_id == disk.Attachments[0]["server_id"] {
+			glog.Infof("Disk: %q is already attached to compute: %q\n", diskName, compute_id)
+			return disk.ID, nil
+		} else {
+			errMsg := fmt.Sprintf("Disk %q is attached to a different compute: %q, should be detached before proceeding", diskName, disk.Attachments[0]["server_id"])
+			glog.Errorf(errMsg)
+			return "", errors.New(errMsg)
+		}
+	}
+	// add read only flag here if possible spothanis
+	_, err = volumeattach.Create(cClient, compute_id, &volumeattach.CreateOpts{
+		VolumeID: disk.ID,
+	}).Extract()
+	if err != nil {
+		glog.Infof("Failed to attach %s volume to %s compute", diskName, compute_id)
+		return "", err
+	}
+	glog.Infof("Successfully attached %s volume to %s compute", diskName, compute_id)
+	return disk.ID, nil
+}
+
+func (os *OpenStack) DetachDisk(partialDiskId string) error {
+	disk, err := os.getVolume(partialDiskId)
+	if err != nil {
+		return err
+	}
+
+	cClient, err := openstack.NewComputeV2(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	compute_id, err := os.getComputeIDbyHostname(cClient)
+	if err != nil || len(compute_id) == 0 {
+		glog.Errorf("Unable to get minion's ip while detaching disk")
+		return err
+	}
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil && compute_id == disk.Attachments[0]["server_id"] {
+		err = volumeattach.Delete(cClient, compute_id, disk.ID).ExtractErr()
```

Comment:
We can leave it as is (maybe put a comment to indicate for future debuggers), if it becomes an issue, we can revisit it.